### ISSUE TITLE
Methods to read/write on plain io.{Reader,Writer}

### ISSUE THREAD

**Issue by gobwas** [gorilla/websocket#186]
Hello!

Thank you for the great library! 🚀 

In application I developing it is very hard restrictions on memory consumption.
So I need to reuse very lot of things, and, of course, buffers.

Is it possible to export some io methods for just to do handshake on plain `net.Conn` or `io.ReadWriter`, and read/write functions on it?

I mean something like this:
```go
conn, err := req.Hijack()
err = websocket.Upgrade(conn)

// eventually
err = websocket.Write(conn, myBytes)

// eventually
bts, err := websocket.Read(conn)
```

**Comment by garyburd** [comment]
Which buffers used by the package are a concern for your application?

**Comment by garyburd** [comment]
I am not in favor of these functions because they make the package more complicated. Please describe the buffers that are a concern for your application and how your application will do things differently.

**Comment by gobwas** [comment]
I think there could not be any complication with optional extracting the buffers from conn. 
I mean the buffers that lay inside `Conn` struct (`writeBuf` and `br`). 
Inside my app Im planning to use epoll to detect the connections that ready to produce some bytes.
Then use buffer from `sync.Pool` to read frames from active connection. 

So pseudocode could look like this:

```go

var epolledConns chan net.Conn
for conn := range epolledConns {
  var buf *bufio.Reader
  buf = acquireBuffer(conn, readBufferSize)

  bts, typ, err := websocket.ReadFrame(buf)
  // ...
}
```

**Comment by gobwas** [comment]
By the way, it is very conformable with (this issue too)[https://github.com/gorilla/websocket/issues/9].

**Comment by garyburd** [comment]
I will not add a feature that requires the app to use epoll directly to make the feature useful. You should copy the package and modify it to meet your needs.

**Comment by gobwas** [comment]
It is not required for the app to use epoll directly. For example, you always know when you want to write data to the socket – so you dont need to keep write buffer for idle connections. With read it also could be useful when you just want to read response after you wrote request.

**Comment by gobwas** [comment]
Alternatively, it is possible just to implement exported constructor of `Conn`, that accepts `net.Conn`, `bufio.Reader`, `bufio.Writer`. What do you think?

**Comment by garyburd** [comment]
>  you just want to read response after you wrote request.

The application must read the connection to (process control messages)[https://godoc.org/github.com/gorilla/websocket#hdr-Control_Messages]. 

> exported constructor of Conn, that accepts net.Conn, bufio.Reader, bufio.Writer

How will be bufio.Reader differ from the one created by the package?  The package does not use a bufio.Writer.

**Comment by gobwas** [comment]
It will not differ, it will be the same, but reused.
Yeah, I realised that there is no buffo.Writer. So, the constructor could accept `bufio.Reader` with `net.Conn` inside, and `[]byte`. This kind of constructor could help to reuse allocated `[]byte`.

**Comment by garyburd** [comment]
New features require a justification and must be useful to more than one application.  

Describe why reusing these buffers will provide a significant performance improvement to your application and how this feature might be useful to other applications.

**Comment by gobwas** [comment]
Reusing these buffers will not increase performance. It will be at most the same. But, it will reduce the memory consumption of application. And this is actual for projects with lots of connections, that almost do nothing periodically.

**Comment by garyburd** [comment]
Are you planning to *share* the buffers between connections?  You will not decrease the memory required by the application if you reuse buffers as connections are opened and closed.

**Comment by gobwas** [comment]
Yes, I am planning to share the buffers between connections. But I will reuse them not on open/close, but on ready to read/write.

**Comment by garyburd** [comment]
It is not possible to share read buffers without use epoll. I will not add a feature that requires the app to use epoll directly to make the feature useful. You should copy the package and modify it to meet your needs.

**Comment by garyburd** [comment]
I have work in progress to fix #9.  You will have complete control of the write buffer when this work is complete.

**Comment by gobwas** [comment]
Sounds useful. Thanks.
